### PR TITLE
TEL-5471: Add prometheus metrics for mod_json_cdr

### PR DESCRIPTION
**Pull request by baloeng**

### CODE REVIEW

File: src/mod/event_handlers/mod_json_cdr/mod_json_cdr.c [hunk 1]
```diff
@@ -305,24 +314,29 @@ static void process_cdr(cdr_data_t *data)
...
 						switch_log_printf(SWITCH_CHANNEL_UUID_LOG(data->uuid), SWITCH_LOG_ERROR, "Error unlinking [%s]\n",path);
 					backup_cdr(data);
 				} else {
+					prometheus_increment_cdr_success();
 					if(!zstr(data->tmpdir)) {
 						char *move_path = switch_mprintf("%s%s%s", data->logdir, SWITCH_PATH_SEPARATOR, data->filename);
 						if(move_path) {
 							if (rename(path, move_path) == 0) {
 								switch_log_printf(SWITCH_CHANNEL_UUID_LOG(data->uuid), SWITCH_LOG_INFO, "Move CDR [%s] to [%s]\n", data->filename, data->logdir);
+								prometheus_increment_tmpcdr_move_success();
```
Comment (dev-ryanc) [inline src/mod/event_handlers/mod_json_cdr/mod_json_cdr.c:328]: Increment both cdr success and tmpcdr move success here

File: src/mod/event_handlers/mod_json_cdr/mod_json_cdr.c [hunk 2]
```diff
@@ -305,24 +314,29 @@ static void process_cdr(cdr_data_t *data)
...
 							} else {
 								// Lets fallback to copy and delete file
 								switch_memory_pool_t *pool = NULL;
 								switch_core_new_memory_pool(&pool);
 								if(pool && switch_file_copy(path, move_path, SWITCH_FPROT_FILE_SOURCE_PERMS, pool) == SWITCH_STATUS_SUCCESS) {
 									switch_log_printf(SWITCH_CHANNEL_UUID_LOG(data->uuid), SWITCH_LOG_INFO, "Move CDR [%s] to [%s]\n", data->filename, data->logdir);
 									switch_file_remove(path, pool);
+									prometheus_increment_tmpcdr_move_success();
```
Comment (dev-ryanc) [inline src/mod/event_handlers/mod_json_cdr/mod_json_cdr.c:336]: Increment both cdr success and tmpcdr move success here

File: src/mod/event_handlers/mod_json_cdr/mod_json_cdr.c [hunk 3]
```diff
@@ -305,24 +314,29 @@ static void process_cdr(cdr_data_t *data)
...
 								} else {
 									switch_log_printf(SWITCH_CHANNEL_UUID_LOG(data->uuid), SWITCH_LOG_ERROR, "Fail to move CDR [%s] to [%s] - %s\n", data->filename, data->logdir, strerror(errno));
+									prometheus_increment_cdr_error();
```
Comment (dev-ryanc) [inline src/mod/event_handlers/mod_json_cdr/mod_json_cdr.c:339]: Increment both cdr error and tmpcdr move failure here

File: src/mod/event_handlers/mod_json_cdr/prometheus_metrics.cpp [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+		_cdr_counter++;
+	}
+
+	void increment_cdr_success()
+	{
+		auto_lock lock(_mutex);
+		_cdr_success++;
+	}
+
+	void increment_tmpcdr_move_success()
```
Comment (dev-ryanc) [inline src/mod/event_handlers/mod_json_cdr/prometheus_metrics.cpp:51]: Add a function tmpcdr move failure

File: src/mod/event_handlers/mod_json_cdr/prometheus_metrics.cpp [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+		stream->write_function(stream, "# TYPE mod_json_cdr_counter counter\n");
+		stream->write_function(stream, "mod_json_cdr_counter %u\n", _cdr_counter);
+
+		stream->write_function(stream, "# HELP mod_json_cdr_success CDRs written to disk\n");
+		stream->write_function(stream, "# TYPE mod_json_cdr_success counter\n");
+		stream->write_function(stream, "mod_json_cdr_success %u\n", _cdr_success);
+
+		stream->write_function(stream, "# HELP mod_json_tmpcdr_success CDRs moved from tmpdir folder\n");
+		stream->write_function(stream, "# TYPE mod_json_tmpcdr_success counter\n");
+		stream->write_function(stream, "mod_json_tmpcdr_success %u\n", _tmpcdr_success);
```
Comment (dev-ryanc) [inline src/mod/event_handlers/mod_json_cdr/prometheus_metrics.cpp:89]: Add a metric for tmpcdr move failure

File: src/mod/event_handlers/mod_json_cdr/prometheus_metrics.cpp [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+{
+	instance->increment_cdr_counter();
+}
+
+void prometheus_increment_cdr_success()
+{
+	instance->increment_cdr_success();
+}
+
+void prometheus_increment_tmpcdr_move_success()
```
Comment (dev-ryanc) [inline src/mod/event_handlers/mod_json_cdr/prometheus_metrics.cpp:153]: Add global function for incrementing tmpcdr move failure

File: src/mod/event_handlers/mod_json_cdr/prometheus_metrics.cpp [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+
+SWITCH_STANDARD_API(json_cdr_prometheus_metrics)
+{
+	instance->generate_metrics(stream);
+	return SWITCH_STATUS_SUCCESS;
+}
+
+SWITCH_BEGIN_EXTERN_C
+
+void prometheus_init(switch_loadable_module_interface_t **module_interface, switch_api_interface_t *api_interface, switch_memory_pool_t* pool)
```
Comment (dev-ryanc) [inline src/mod/event_handlers/mod_json_cdr/prometheus_metrics.cpp:129]: I didn't see you calling this method on load init

File: src/mod/event_handlers/mod_json_cdr/prometheus_metrics.cpp [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+{
+	SWITCH_ADD_API(api_interface, "json_cdr_prometheus_metrics", "json_cdr_prometheus_metrics", json_cdr_prometheus_metrics, "");
+	
+	delete instance;
+	instance = new prometheus_metrics(module_interface, pool);
+}
+
+void prometheus_destroy()
```
Comment (dev-ryanc) [inline src/mod/event_handlers/mod_json_cdr/prometheus_metrics.cpp:137]: I didn't see this calling this method on shutdown

**Review by dev-ryanc — CHANGES_REQUESTED** [review]

File: src/mod/event_handlers/mod_json_cdr/mod_json_cdr.c [hunk 9]
```diff
@@ -305,24 +314,29 @@ static void process_cdr(cdr_data_t *data)
 				close(fd);
 				if (x < 0) {
 					switch_log_printf(SWITCH_CHANNEL_UUID_LOG(data->uuid), SWITCH_LOG_ERROR, "Error writing [%s]\n",path);
+					prometheus_increment_cdr_error();
 					if (0 > unlink(path))
 						switch_log_printf(SWITCH_CHANNEL_UUID_LOG(data->uuid), SWITCH_LOG_ERROR, "Error unlinking [%s]\n",path);
 					backup_cdr(data);
 				} else {
+					prometheus_increment_cdr_success();
```
[the remark this answers is not in the document]
Comment (baloeng) [reply src/mod/event_handlers/mod_json_cdr/mod_json_cdr.c]: Done

File: src/mod/event_handlers/mod_json_cdr/mod_json_cdr.c (hunk 1 above)
Comment (baloeng) [reply src/mod/event_handlers/mod_json_cdr/mod_json_cdr.c:328]: Done

File: src/mod/event_handlers/mod_json_cdr/mod_json_cdr.c (hunk 2 above)
Comment (baloeng) [reply src/mod/event_handlers/mod_json_cdr/mod_json_cdr.c:337]: Done

File: src/mod/event_handlers/mod_json_cdr/mod_json_cdr.c (hunk 3 above)
Comment (baloeng) [reply src/mod/event_handlers/mod_json_cdr/mod_json_cdr.c:340]: Done

File: src/mod/event_handlers/mod_json_cdr/prometheus_metrics.cpp (hunk 4 above)
Comment (baloeng) [reply src/mod/event_handlers/mod_json_cdr/prometheus_metrics.cpp:58]: Done

File: src/mod/event_handlers/mod_json_cdr/prometheus_metrics.cpp (hunk 5 above)
Comment (baloeng) [reply src/mod/event_handlers/mod_json_cdr/prometheus_metrics.cpp:100]: Done

File: src/mod/event_handlers/mod_json_cdr/prometheus_metrics.cpp (hunk 7 above)
Comment (baloeng) [reply src/mod/event_handlers/mod_json_cdr/prometheus_metrics.cpp]: Done

File: src/mod/event_handlers/mod_json_cdr/prometheus_metrics.cpp (hunk 8 above)
Comment (baloeng) [reply src/mod/event_handlers/mod_json_cdr/prometheus_metrics.cpp:150]: Done

File: src/mod/event_handlers/mod_json_cdr/prometheus_metrics.cpp (hunk 6 above)
Comment (baloeng) [reply src/mod/event_handlers/mod_json_cdr/prometheus_metrics.cpp:171]: Done

**Review by dev-ryanc — CHANGES_REQUESTED** [review]
You need to add a else statement in line 322 to increment cdr success
```
				if(!zstr(data->tmpdir)) {
						char *move_path = switch_mprintf("%s%s%s", data->logdir, SWITCH_PATH_SEPARATOR, data->filename);
```
This means that if tmpcdr is disable then we already generate the cdr file to the main folder. We need to count that as a success.

**Comment by baloeng** [comment]
I actually added the else statement, forgot to commit it...my bad

**Comment by baloeng** [comment]
Done

**Pull request merged**